### DISCUSSION TITLE
Can you namespace metadata?

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#8954] [Q&A] [answered]
### Description

Is it possible to write a filter that allows the of reuse metadata keys without hiding them in under distinct keys? I'm 95% sure the answer is no, but I'm curious about others' thoughts.

What I'd like to do is have a document that looks like this:

```
---
title: "My Doc"
---

## {#g038py7}

---
title: Favorite Color
---

What is your favorite color?

## {#y293xa8}

---
title: Favorite Food
---

What is your favorite food?
```

That, effectively, becomes this document though a filter:

```
---
title: "My Doc"
---

## {#g038py7}

---
g038py7:
  title: Favorite Color
---

What is your favorite color?

## {#y293xa8}

---
y293xa8:
  title: Favorite Food
---

What is your favorite food?
```

The pandoc manual says what'll happen in the first version:

> A document may contain multiple metadata blocks. If two metadata blocks attempt to set the same field, the value from the second block will be taken.

So first version will only have one `title` metadata key in the AST with a value of `My Doc` whereas the second version will have both that meta as well as `g038py7: title: Favorite Color` and `y293xa8: title: Favorite Food`.

The challenge is that by the time the filter runs (even `pre-ast`), the metadata merging will have already happened, so it's not possible to recover the other metadata blocks. This is just how pandoc works and so maybe that's the end of the story. What struck me, though, is that this is different from how cell execution options work. It's no problem to have one cell be `echo: true` and the next be `echo: false`; there is no global cell option merging (though you can set a global default).

Maybe one reason it's so natural to namespace the execution options within a cell is because the content of that cell is well-defined (all the stuff inside the three ticks). There is, though, an existing example of similarly well-defined blocks of content: revealjs slide sections. Right now, you specify slide-level meta data using pandocs attribute syntax on the headers. That's always felt a bit kludgy to me: they look more like Rmd chunk options than Quarto hashpipe options and you can't nest them.

Is it sensible to think of a pre-pandoc parser that would grab ids from level 2 headers and use them to namespace any metadata found within that section?

**Comment by mcanouil** [comment]
Filters happen on the AST.
Metadata have already been gathered into "metadata" inside the AST.

If you want to do something like that, you need to go to a lower level and implement a Pandoc reader (basically a parser), but even if you do this, then you'll have to provide the results to Quarto.

This being said, I don't understand the purpose of what you are trying to do.

**Reply by andrewpbray** [reply]
Yep, a pandoc parser would certainly do the trick but I figured I'd raise the issue here at Quarto because Quarto allow cell-specific metadata via the hashpipe and there's a parallel between that are section-specific metadata.

The immediate use case I'm looking at is writing a quarto quiz custom format. The document consists of a series of questions, each of which have their own metadata: title, tags, and variables, used for creating variants of the question for different versions. I'd like it to be possible to author questions either directly in the body of the quiz document or in separate files that are then included in the quiz document, and for the metadata syntax to look as similar as possible.

**Comment by cscheid** [comment, marked as answer]
Unfortunately no. Fundamentally, the problem is that Pandoc coalesces all metadata blocks before we get the chance to do anything with them. At the surface level, we could come up with a system that does "block-level metadata", something like

````
```{quarto-metadata}
Your yaml here
```
````

and which obeys different resolution rules, but this would be semantically very hard to teach. For example: what happens if conditional content blocks have `{quarto-metadata}` inside them which changes the values of the metadata that other conditional blocks access? What's the order of resolution? Etc.

With that said, your idea of block-scoping metadata does looks implementable with the syntax I proposed above. If it's done by you, then all the semantic complication is not a problem (you built it, you should presumably know what you're doing with it).

**Reply by cscheid** [reply]
Hopefully the way the filter works is relatively straightforward to see. You'd write a `Pandoc` entry that walks the headings (or use the blocks-to-sections Pandoc API whose name escapes me) in order, finds the code blocks with `.andrew-bray-metadata` or whatever classes, parses them as YAML (you'd have to find a pure-lua YAML parser. Fun!) and then adds them to the document metadata in the right scope.

**Reply by andrewpbray** [reply]
Haha, I was thinking about that. Then I thought about

> you'd have to find a pure-lua YAML parser.

We'll see.

The motivation for this particular format comes from a colleague. I have a v.0000001 of this quiz format [here](https://github.com/stat20/stat20quiz?tab=readme-ov-file#make-versions) that implements a leaner-syntax version of conditional content for spans to do the versioning.

```markdown
The *mean* is a measure of [spread]{.vA}[center]{.vB}.
```
My colleague, a new quarto user, found this syntax hard to read and expected it to see it as yaml metadata, something like:

```
---
mean-measure:
  A: spread
  B: center
___

The *mean* is a measure of {{@ mean-measure @}}.
```
The more I think about it, the more I like his version.

So... we'll see. Maybe I'll take a big breath some day and look into integrating one of the [existing lua parsers](https://github.com/exosite/lua-yaml)...

**Reply by cscheid** [reply]
Your syntax version seems possible doable if you switch to shortcodes:

````
---
mean-measure:
  A: spread
  B: center
---

The *mean* is a measure of {{< quiz mean-measure >}}.
````
